### PR TITLE
Custom and global contexts

### LINKED ISSUE #24: Attach contexts to events
Snowplow allows to [attach contexts to events](https://docs.snowplowanalytics.com/docs/collecting-data/collecting-from-own-applications/mobile-trackers/previous-versions/objective-c-tracker/ios-tracker-1-7-0/#attaching-contexts-to-events) by calling `.setContext` on them. We could leverage `AbstractEvent` and add a `.setContexts` method OR add contexts as a parameter to AbstractEvent's initialiser and all events that implement it. 

Problem with the AbstractEvent init approach is that we would limit ourselves to Snowplow events and will not be able to extend this library's functionality with other trackers in the future if we wanted to. We could add another layer of inheritance `AbstractSnowplowEvent` that then allows to add contexts and is an `AbstractEvent`.

We could also add an additional AbstractTracker that enriches any `AbstractSnowplowEvent` with a given `() => Context`.

### PR DESCRIPTION
**Pull request by ohitsdaniel**
Resolves #24.

Allows to attach custom contexts to events. Also adds GlobalContextProvider which attaches contexts to all events that pass through it.

### CODE REVIEW

File: lib/src/tracker/trackers/tracking_guard.dart [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+class TrackingGuard extends AbstractTracker {
+  final AbstractTracker _child;
+  final Future<bool> Function(AbstractEvent) _shouldTrack;
+
+  /// Default initialiser
+  const TrackingGuard({
+    required AbstractTracker child,
+    required Future<bool> Function(AbstractEvent) shouldTrack,
+  })   : _child = child,
+        _shouldTrack = shouldTrack;
```
Comment (MisterJimson) [inline lib/src/tracker/trackers/tracking_guard.dart:18]: Minor nit: I prefer using the `this.` syntax for populating properties in the constructor. It means the properties need to be public, but that seems fine here, they are final anyway.

**Review by MisterJimson — CHANGES_REQUESTED** [review]
1 very minor nit, if you have a strong reason the fields should remain private let me know.

File: lib/src/tracker/trackers/tracking_guard.dart (hunk 1 above)
Comment (ohitsdaniel) [reply lib/src/tracker/trackers/tracking_guard.dart:18]: Guess it's just something that stuck with me from Swift development, where it is good practice to keep all fields private, unless they are 'needed' publicly. This avoids misuse as library users won't be able to circumvent the guard by calling tracker.child.track(..) directly.

Comment (MisterJimson) [reply lib/src/tracker/trackers/tracking_guard.dart:18]: Sounds good, yeah I get that. If there is a decent reason to I am fine to keep.

**Review by MisterJimson — APPROVED** [review]

**Pull request merged**